File: apps/android/feature/ai/data/src/main/java/com/scribblefit/feature/ai/data/repository/ConfigRepositoryImpl.kt
package com.scribblefit.feature.ai.data.repository

import com.scribblefit.core.database.dao.SystemConfigDao
import com.scribblefit.core.database.entity.SystemConfigEntity
import com.scribblefit.feature.ai.domain.engine.ConfigRepository
import com.scribblefit.feature.ai.domain.model.LLMProvider
import com.scribblefit.feature.ai.domain.model.SystemConfig
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ConfigRepositoryImpl @Inject constructor(
    private val systemConfigDao: SystemConfigDao
) : ConfigRepository {

    override fun getConfig(): Flow<SystemConfig?> =
        systemConfigDao.observe().map { it?.toDomain() }

    override suspend fun updateConfig(config: SystemConfig) {
        systemConfigDao.upsert(config.toEntity())
    }

    private fun SystemConfigEntity.toDomain(): SystemConfig = SystemConfig(
        promptVersion = promptVersion,
        promptText = promptText,
        exerciseVersion = exerciseVersion,
        preferredLlmProvider = LLMProvider.entries.find { it.rawValue == preferredLlmProvider } ?: LLMProvider.PROXY,
        preferredModel = preferredModel,
        parsingMode = parsingMode,
        weightUnit = weightUnit,
        themePreference = themePreference,
        updatedAt = updatedAt
    )

    private fun SystemConfig.toEntity(): SystemConfigEntity = SystemConfigEntity(
        id = "config",
        promptVersion = promptVersion,
        promptText = promptText,
        exerciseVersion = exerciseVersion,
        preferredLlmProvider = preferredLlmProvider.rawValue,
        preferredModel = preferredModel,
        parsingMode = parsingMode,
        weightUnit = weightUnit,
        themePreference = themePreference,
        updatedAt = updatedAt
    )
}

File: apps/ios/ScribbleFit/Sources/Features/AI/Data/Repository/ConfigRepositoryImpl.swift
import Foundation

@MainActor
public final class ConfigRepositoryImpl: ConfigRepository {
    private let database: ScribbleFitDatabase

    public init(database: ScribbleFitDatabase) {
        self.database = database
    }

    public func getConfig() async -> SystemConfigDomain? {
        database.getConfig().map { entity in
            SystemConfigDomain(
                promptVersion: entity.promptVersion,
                promptText: entity.promptText,
                exerciseVersion: entity.exerciseVersion,
                preferredLlmProvider: LLMProvider(rawValue: entity.preferredLlmProvider) ?? .proxy,
                preferredModel: entity.preferredModel,
                parsingMode: entity.parsingMode,
                weightUnit: entity.weightUnit,
                themePreference: entity.themePreference,
                updatedAt: entity.updatedAt
            )
        }
    }

    public func updateConfig(_ config: SystemConfigDomain) async throws {
        database.upsertConfig(SystemConfig(
            id: "config",
            promptVersion: config.promptVersion,
            promptText: config.promptText,
            exerciseVersion: config.exerciseVersion,
            preferredLlmProvider: config.preferredLlmProvider.rawValue,
            preferredModel: config.preferredModel,
            parsingMode: config.parsingMode,
            weightUnit: config.weightUnit,
            themePreference: config.themePreference,
            updatedAt: config.updatedAt
        ))
    }
}
